pilot/pkg/networking/grpcgen/grpc.go

```diff
@@ -0,0 +1,273 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package grpcgen
+
+import (
+	"net"
+	"strconv"
+	"strings"
+
+	xdsapi "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	envoycore "github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+	envoy_api_v2_route "github.com/envoyproxy/go-control-plane/envoy/api/v2/route"
+	v2 "github.com/envoyproxy/go-control-plane/envoy/config/filter/network/http_connection_manager/v2"
+	envoy_config_listener_v2 "github.com/envoyproxy/go-control-plane/envoy/config/listener/v2"
+	gogo "github.com/gogo/protobuf/proto"
+	"github.com/golang/protobuf/ptypes/any"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/networking/util"
+	"istio.io/istio/pkg/config/schema/resource"
+	"istio.io/pkg/log"
+
+	"istio.io/istio/pkg/config/host"
+)
+
+// To avoid a recoursive depenency to v2.
+const (
+	typePrefix = "type.googleapis.com/envoy.api.v2."
+
+	// Constants used for XDS
+
+	// ClusterType is used for cluster discovery. Typically first request received
+	ClusterType = typePrefix + "Cluster"
+	// EndpointType is used for EDS and ADS endpoint discovery. Typically second request.
+	EndpointType = typePrefix + "ClusterLoadAssignment"
+	// ListenerType is sent after clusters and endpoints.
+	ListenerType = typePrefix + "Listener"
+	// RouteType is sent after listeners.
+	RouteType = typePrefix + "RouteConfiguration"
+)
+
+// Support generation of 'ApiListener' LDS responses, used for native support of gRPC.
+// The same response can also be used by other apps using XDS directly.
+
+// GRPC proposal:
+// https://github.com/grpc/proposal/blob/master/A27-xds-global-load-balancing.md
+//
+// Note that this implementation is tested against gRPC, but it is generic - any other framework can
+// use this XDS mode to get load balancing info from Istio, including MC/VM/etc.
+
+// DNS can populate the name to cluster VIP mapping using this response.
+
+// The corresponding RDS response is also generated - currently gRPC has special differences
+// and can't understand normal Istio RDS - in particular expects "" instead of "/" as
+// default prefix, and is expects just the route for one host.
+// handleAck will detect if the message is an ACK or NACK, and update/log/count
+// using the generic structures. "Classical" CDS/LDS/RDS/EDS use separate logic -
+// this is used for the API-based LDS and generic messages.
+type GrpcConfigGenerator struct {
+}
+
+func (g *GrpcConfigGenerator) Generate(node *model.Proxy, push *model.PushContext, w *model.WatchedResource) []*any.Any {
+	switch w.TypeUrl {
+	case ListenerType:
+		return g.BuildListeners(node, push, w.ResourceNames)
+	case ClusterType:
+		return g.BuildClusters(node, push, w.ResourceNames)
+	case RouteType:
+		return g.BuildHTTPRoutes(node, push, w.ResourceNames)
+	default:
+		return g.handleConfigResource(node, push, w)
```

Comment:
It seems misplaced to have this under grpc since they are not using it (?)

Reply:
I consider gRPC to be the first user - I'll rename the directory to make it clear.
I actually started this work to allow DNS to pull the ApiListeners - just realized that gRPC was using the same thing and was easier to test with their implementation.

And finally the 'api' mode is part of prototype to move generation in agent - may or may not work, but we can decide based on data. As I mentioned I'll keep this PR as an active branch - and gradually move stuff that is ready via separate PRs.